google/stackdriver_monitoring/stackdriver_handlers.py

```diff
@@ -98,78 +187,206 @@ def __init__(self, options, stackdriver):
+      logging.info('batch CREATE %s', name)
+      return create_method(
+          name='projects/{0}'.format(project), body=descriptor)
+
+    create_processor = BatchProcessor(
+        project, self.__stackdriver,
+        create_list, create_invocation, get_descriptor_name)
+    create_processor.process()
+
+    for index, ok in enumerate(create_processor.was_ok):
+      if ok:
+        success_list.append(create_list[index])
+      else:
+        failed_list.append(create_list[index])
+        failed_errors.append(create_processor.batch_response[index])
+
+  def __do_batch_update(self, project, update_list, original_type_map):
+    get_descriptor_name = lambda descriptor: descriptor['name']
+
+    delete_errors = []
+    create_errors = []
+    restore_errors = []
+
+    failed_list = []
+    create_list = []
+    success_list = []
+    restore_list = []
+    not_updated_list = []
+    lost_list = []
+
+    if update_list:
+      self.__do_batch_update_delete_helper(
+          project, update_list, create_list, failed_list, delete_errors)
+
+    if create_list:
+      self.__do_batch_update_create_helper(
+        project, create_list, success_list, restore_list, create_errors)
+      restore_list = [original_type_map[elem['type']] for elem in restore_list]
+
+    if restore_list:
+      self.__do_batch_update_create_helper(
+        project, restore_list, not_updated_list, lost_list, restore_errors)
+
+    response_code = 200 if len(failed_list) == 0 else 500
+    bodies = []
+    for elem in success_list:
+      bodies.append('Updated {0} to {1}'.format(elem['type'], elem))
+    for index, elem in enumerate(failed_list):
+      bodies.append('Failed to update {0} to {1}: {2}'.format(
+        elem['type'], elem, delete_errors[index]))
+    for index, elem in enumerate(restore_list):
+      bodies.append('Failed to update {0} to {1}: {2}'.format(
+        elem['type'], elem, create_errors[index]))
+    for index, elem in enumerate(lost_list):
+      bodies.append('Lost {0}. It used to be {1}: {2}'.format(
+        elem['type'], elem, restore_errors[index]))
+
+    return response_code, {'Content-Type': 'text/plain'}, '\n'.join(bodies)
+
+  def upsert_descriptors(
+      self, project, upsert_descriptors, type_map, response_collector):
+    create_list = []
+    update_list = []
+    for elem in upsert_descriptors:
+      elem_type = elem.get('type', '')
+      if not elem_type.startswith('custom.googleapis.com/spinnaker'):
+        raise ValueError('Invalid Metric Descriptor:\n{0}\n'.format(elem))
+      if elem_type in type_map:
+        if elem != type_map[elem_type]:
+          update_list.append(elem)
+      else:
+        create_list.append(elem)
+
+    headers = {}
+    response_code = 200
+    response_body = []
+
+    if create_list:
+      create_response_code, create_headers, create_response_body = (
+          self.__do_batch_create(project, create_list))
+      response_code = max(response_code, create_response_code)
```

Comment:
Agreed. I'm not sure what to do here and other parts need cleaned up as well so saving it for a rainy day to go back over.